google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectLoader.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.project;
+
+import com.google.api.client.repackaged.com.google.common.base.Strings;
+import com.google.api.services.cloudresourcemanager.CloudResourceManager;
+import com.google.api.services.cloudresourcemanager.model.ListProjectsResponse;
+import com.google.api.services.cloudresourcemanager.model.Project;
+import com.google.cloud.tools.intellij.login.CredentialedUser;
+import com.google.cloud.tools.intellij.resources.GoogleApiClientFactory;
+import com.google.cloud.tools.intellij.util.GctBundle;
+import com.google.cloud.tools.intellij.util.ThreadUtil;
+import com.google.common.annotations.VisibleForTesting;
+import com.intellij.openapi.diagnostic.Logger;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+
+/** Loads list of {@link Project} for a {@link CredentialedUser}. */
+class ProjectLoader {
+  private static final Logger LOG = Logger.getInstance(ProjectLoader.class);
+
+  private static final int PROJECTS_MAX_PAGE_SIZE = 300;
+  private static final String PROJECT_DELETE_REQUESTED = "DELETE_REQUESTED";
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  void loadUserProjectsInBackground(
```

Comment:
Do you think the callback would be better represented using a future? In fact, ThreadUtil#executeInBackground does return a future.

Reply:
I think standard Future is too simplistic for us - it does not allow notification when task is done, and we can't block UI. Do you mean Guava listenable future like https://github.com/google/guava/wiki/ListenableFutureExplained? From initial view it would work. Only question is will it be a bit harder to read code that deal with generic future rather than specific callback interface, but it's probably minor.
If you think it makes sense, I can convert it.

Reply:
Technically you can add a listener to the returned `Future`, but it's not as great as `ListenableFuture`, hence why it was created in the first place.

I support changing the `ThreadUtil` class to use a listening decorator on its cached thread pool. The only reason I didn't do this is because up until now, we had no need for the returned future. Now that we do, it makes sense to provide a better interface to it via `ListenableFuture`.

I can send a PR if you want, or you can do it in this PR. It's up to you.

Reply:
Just to point out. JDK 8 has the `CompletableFuture` which to my knowledge can accomplish the same thing. @nkibler any idea which is currently preferred internally?

Reply:
Seems like `ListenableFuture` is still the preferred type: http://go/java8walkthrough/redundancy#listenablefuture-vs-completablefuture

Reply:
ah go/java8walkthrough/redundancy. pretty ambiguous, but seems to be a nod towards `ListenableFuture`

Reply:
It should be a couple of line change in ThreadUtil, I think it makes sense to add it in this PR.